bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogManager.java

```diff
@@ -0,0 +1,240 @@
+ * under the License.
+ *
+ */
+
+package org.apache.bookkeeper.bookie;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.util.concurrent.FastThreadLocal;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.bookkeeper.bookie.EntryLogger.BufferedLogChannel;
+import org.apache.bookkeeper.bookie.EntryLogger.EntryLogListener;
+import org.apache.bookkeeper.bookie.LedgerDirsManager.NoWritableLedgerDirException;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+
+interface EntryLogManager {
+
+    /*
+     * add entry to the corresponding entrylog and return the position of
+     * the entry in the entrylog
+     */
+    long addEntry(long ledger, ByteBuf entry, boolean rollLog) throws IOException;
+
+    /*
+     * gets the active logChannel with the given entryLogId. null if it is
+     * not existing.
+     */
+    BufferedLogChannel getCurrentLogIfPresent(long entryLogId);
+
+    /*
+     * Returns eligible writable ledger dir for the creation next entrylog
+     */
+    File getDirForNextEntryLog(List<File> writableLedgerDirs);
+
+    /*
+     * Do the operations required for checkpoint.
+     */
+    void checkpoint() throws IOException;
+
+    /*
+     * flush both current and rotated logs.
+     */
+    void flush() throws IOException;
+
+    /*
+     * close current logs.
+     */
+    void close() throws IOException;
+
+    /*
+     * force close current logs.
+     */
+    void forceClose();
+
+    /*
+     * prepare entrylogger/entrylogmanager before doing SortedLedgerStorage
+     * Checkpoint.
+     */
+    void prepareSortedLedgerStorageCheckpoint(long numBytesFlushed) throws IOException;
+
+    /*
+     * this method should be called before doing entrymemtable flush, it
+     * would save the state of the entrylogger before entrymemtable flush
+     * and commitEntryMemTableFlush would take appropriate action after
+     * entrymemtable flush.
+     */
+    void prepareEntryMemTableFlush();
+
+    /*
+     * this method should be called after doing entrymemtable flush,it would
+     * take appropriate action after entrymemtable flush depending on the
+     * current state of the entrylogger and the state of the entrylogger
+     * during prepareEntryMemTableFlush.
+     *
+     * It is assumed that there would be corresponding
+     * prepareEntryMemTableFlush for every commitEntryMemTableFlush and both
+     * would be called from the same thread.
+     *
+     * returns boolean value indicating whether EntryMemTable should do checkpoint
+     * after this commit method.
+     */
+    boolean commitEntryMemTableFlush() throws IOException;
+
+    /*
+     * creates new separate log for compaction.
+     */
+    BufferedLogChannel createNewLogForCompaction() throws IOException;
+
+    @Slf4j
+    abstract class EntryLogManagerBase implements EntryLogManager {
```

Comment:
better move `EntryLogManagerBase` to a separate file as well, to make this a pure interface file.

Reply:
will do